src/OpenTelemetry/Metrics/Export/PushMetricController.cs

```diff
@@ -0,0 +1,122 @@
+                {
+                    var sw = Stopwatch.StartNew();
+
+                    foreach (var meter in this.meters.Values)
+                    {
+                        (meter as MeterSdk).Collect();
+                    }
+
+                    OpenTelemetrySdkEventSource.Log.CollectionCompleted(sw.ElapsedMilliseconds);
+
+                    // Collection is over at this point. All metrics are given
+                    // to the MetricProcesor(Batcher).
+                    // Let MetricProcessor know that this cycle is ending,
+                    // and send the metrics from MetricProcessor
+                    // to the MetricExporter.
+                    var metricsToExportTuple = this.metricProcessor.FinishCollectionCycle();
+                    if (metricsToExportTuple.Item1 != null)
+                    {
+                        longMetricToExport.AddRange(metricsToExportTuple.Item1);
+                    }
+
+                    if (metricsToExportTuple.Item2 != null)
+                    {
+                        doubleMetricToExport.AddRange(metricsToExportTuple.Item2);
+                    }
+
+                    var longExportResult = await this.metricExporter.ExportAsync<long>(longMetricToExport, cancellationToken);
+                    if (longExportResult != MetricExporter.ExportResult.Success)
+                    {
+                        OpenTelemetrySdkEventSource.Log.MetricExporterErrorResult(longExportResult);
+
+                        // we do not support retries for now and leave it up to exporter
+                        // as only exporter implementation knows how to retry: which items failed
+                        // and what is the reasonable policy for that exporter.
+                    }
+
+                    longMetricToExport.Clear();
+
+                    var doubleExportResult = await this.metricExporter.ExportAsync<double>(doubleMetricToExport, cancellationToken);
+                    if (doubleExportResult != MetricExporter.ExportResult.Success)
+                    {
+                        OpenTelemetrySdkEventSource.Log.MetricExporterErrorResult(longExportResult);
+
+                        // we do not support retries for now and leave it up to exporter
+                        // as only exporter implementation knows how to retry: which items failed
+                        // and what is the reasonable policy for that exporter.
+                    }
+
+                    doubleMetricToExport.Clear();
+
+                    var remainingWait = this.pushInterval - sw.Elapsed;
+                    if (remainingWait > TimeSpan.Zero)
+                    {
+                        await Task.Delay(remainingWait, cancellationToken).ConfigureAwait(false);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                OpenTelemetrySdkEventSource.Log.MetricControllerException(ex);
```

Comment:
so a single exception will stop exporter completely? Perhaps we can have another inner `try{}catch`?

Reply:
Good catch. Yes will fix this .